---
title: "`scAnno` package vignette"
author:"Liu hongjia"
date:"`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{scAnno-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(scAnno)
```

# Introduction
Undoubtedly, single-cell RNA sequencing (scRNA-seq) has changed the research landscape by providing insights into heterogeneous, complex and rare cell populations. Given that more such datasets will become available in the near future, their accurate assessment with compatible and robust models for cell type annotation is a prerequisite.scAnno is an automated annotation tool for single-cell RNA sequencing datasets primarily based on the single cell cluster levels, using a joint deconvolution strategy and logistic regression.

# Application of `scAnno`
First, we should load the `scAnno` package.

```{r load_libraries,  message=FALSE, results='hide', warning=FALSE}

library(scAnno)
```
We explicitly created a complete reference atlas (reference expression profiles) of 30 cell types from the Human Cell Landscape (HCL) covering more than 50 human tissues to support scAnno.scAnno offers a possibility to obtain genes with high expression and specificity in a given cell type as cell type-specific genes (marker genes) by combining co-expression genes with seed genes as a core.

#Import human cell type reference profile.
data(Human_cell_landscape)

#Import protein coding gene(19814 genes) to filter reference expression profile.
data(gene.anno)

#Import TCGA bulk data in pan-cancer.
data(tcga.data.u)

scAnno requires a object data which is processed by Seurat standard process, here we use GSE136103 dataset as an example to show the annotation. 
#A liver tissue data set to be annotated.
data(GSE136103)

#Seurat object, which need to be annotated.
obj.seu <- GSE136103

#Seurat object of reference gene expression profile.
ref.obj <- Human_cell_landscape

#Reference gene expression profile.
ref.expr <- GetAssayData(ref.obj, slot = 'data') %>% as.data.frame

#Cell type information of reference profile, corresponding to the above `ref.expr`.
ref.anno <- Idents(ref.obj) %>% as.character

After preparing the data, we start running scAnno.

    scAnno(query = obj.seu,
    ref.expr = ref.expr,
    ref.anno = ref.anno ,
    save.markers = "ref.markers",
    cluster.col = "seurat_clusters",
    factor.size = 0.1,
    pvalue.cut = 0.01,
    seed.num = 10,
    redo.markers = FALSE,
    gene.anno = gene.anno,
    permut.num = 100,
    show.plot = TRUE,
    verbose = TRUE,
    tcga.data.u = tcga.data.u
    )













