
---
title: "`scAnno` package vignette"
author:"Hongjia Liu"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{scAnno-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r}
library(scAnno)
```

# Introduction
Undoubtedly, single-cell RNA sequencing (scRNA-seq) has changed the research landscape by providing insights into heterogeneous, complex and rare cell populations. Given that more such datasets will become available in the near future, their accurate assessment with compatible and robust models for cell type annotation is a prerequisite. Considering this, herein, we developed `scAnno` (scRNA-seq data annotation), an automated annotation tool for single-cell RNA sequencing datasets primarily based on the single cell cluster levels, using a joint deconvolution strategy and logistic regression. We explicitly created complete human reference atlas (30 cell types from the Human Cell Landscape covering more than 50 tissues) and mouse reference atlas (26 cell types from Mouse Cell Atlas covering nearly 50 tissues) to support this novel methodology (scAnno).The human single cell reference profile (hcl.sc.rda) and  the mouse single cell reference profile (mca.sc.rda) are  built into scAnno.Users can import appropriate reference  expression profile according to species. Built-in data includes list of protein-coding genes(gene.anno),TCGA bulk expression data after dimensionality reduction(tcga.data.u),and a liver tissue dataset for test(GSE136103).


# Dependencies
scAnno requires a dependency on R versions larger than 3.5.0, in which python is used interactively, and requires the installation of the sklearn module.The dependent packages are Seurat, dplyr, reticulate, MASS, irlba, future, progress, parallel, glmnet, knitr, rmarkdown, devtools.

# Application
```{r}
#Installing scAnno using github
devtools::install_github("liuhong-jia/scAnno")  
library(scAnno)

#Import human cell type reference profile.
data(hcl.sc)

#Import protein coding gene(19814 genes) to filter reference expression profile.
data(gene.anno)

#Import TCGA bulk data in pan-cancer.
data(tcga.data.u)

#A liver tissue data set to be annotated.
data(GSE136103)
```

# Preparing data for input
```{r}
# Seurat object, which need to be annotated.
obj.seu <- GSE136103

#Seurat object of reference gene expression profile.
ref.obj <- hcl.sc

#Reference gene expression profile.
ref.expr <- GetAssayData(ref.obj, slot = 'data') %>% as.data.frame

#Cell type information of reference profile, corresponding to the above `ref.expr`.
ref.anno <- Idents(ref.obj) %>% as.character
```
# scRNA-seq data annotation 
```{r}
results = scAnno(query = obj.seu,
	  ref.expr = ref.expr,
	  ref.anno = ref.anno,
	  save.markers = "markers",
	  cluster.col = "seurat_clusters",
	  factor.size = 0.1,
	  pvalue.cut = 0.01,
	  seed.num = 10,
	  redo.markers = FALSE,
	  gene.anno = gene.anno,
	  permut.num = 100,
	  permut.p = 0.01,
	  show.plot = TRUE,
	  verbose = TRUE,
	  tcga.data.u = tcga.data.u
	  )
 ```
# Results
`results` contains pred.label,`pred.score` and `pvals`.`pred.label` indicates cell types corresponding to each cluster,`pred.score` indicates prediction score for each cluster,corresponding to `pred.label`,`pvals` indicates significance level of the predicted scores, corresponding to `pred.score`.

```{r}
results$query
```
 
```{r}
results$reference
```

```{r}
results$pred.score
```

```{r}
results$pvals
```
# Visualization

```{r}
DimPlot(results$query, group.by = "seurat_clusters", label = TRUE, label.size = 6) | DimPlot(results$query, group.by = 'scAnno', label = TRUE,label.size = 6)
```
 


